packages/pcd/semaphore-identity-v4/src/SemaphoreIdentityV4PCDPackage.ts

```diff
@@ -0,0 +1,107 @@
+import { DisplayOptions, PCDPackage, SerializedPCD } from "@pcd/pcd-types";
+import { encodePrivateKey, encodePublicKey } from "@pcd/pod";
+import { SemaphoreIdentityPCD } from "@pcd/semaphore-identity-pcd";
+import { generateSnarkMessageHash, requireDefinedParameter } from "@pcd/util";
+import { Identity } from "@semaphore-protocol/identity";
+import { derivePublicKey } from "@zk-kit/eddsa-poseidon";
+import JSONBig from "json-bigint";
+import { v4 as uuid } from "uuid";
+import {
+  SemaphoreIdentityV4PCD,
+  SemaphoreIdentityV4PCDArgs,
+  SemaphoreIdentityV4PCDClaim,
+  SemaphoreIdentityV4PCDProof,
+  SemaphoreIdentityV4PCDTypeName
+} from "./SemaphoreIdentityV4PCD";
+
+export function v4PublicKey(identity: Identity): string {
+  const unpackedPublicKey = derivePublicKey(
+    Buffer.from(identity.export(), "base64")
+  );
+  const publicKey = encodePublicKey(unpackedPublicKey);
+  return publicKey;
+}
+
+export function v4PrivateKey(identity: Identity): string {
+  return encodePrivateKey(Buffer.from(identity.export(), "base64"));
+}
+
+export function v3tov4Identity(
+  v3Identity: SemaphoreIdentityPCD
+): SemaphoreIdentityV4PCD {
+  const derivedPrivateKey = Buffer.from(
```

Comment:
Drive-by comment based on our offline discussion:

This works but has extra conversion steps which are quirky and specific to our code.  generateSnarkMessageHash is built for the limitations of field elements, which is why it outputs only 31 bytes not 32.

SV4 private keys can be anything, but it's most convenient if they're 32 bytes of binary (because that's what PODs have normalized on).  I'd recommend we make this transformation something well-specified at the byte level so it's easy to reproduce in other languages if necessary.  Something like this:

```ts
const hashInput = Buffer.from(beBigIntToBuffer(trapdoor, 32).toString("hex") + beBigIntToBuffer(nullifier, 32).toString("hex"), "hex");
const privKey = Buffer.from(sha256(hashInput), "hex");
```

All the hex-conversions are just for code readibility, but in English this is just converting each part of the secret to 32 bytes big-endian, appending them together, hashing using sha256, and using the resulting 32 bytes as the private key.

I'm open to alternatives or tweaks.  Big vs. little-endian in particular feels arbitrary to me, I'm not sure there's a "right" answer.  (zk-kit defaults to BE, iden3 libraries like ffjavascript all seem to prefer LE). The important think to me is that we're explicitly specifying all the details (like byte lengths and endianness) so we're sure we know how to reproduce the same mechanism in another language, and we don't get quirky behavior if the random inputs happen to be small enough to fit into fewer than the max number of bytes.

Reply:
Sounds good to me, thank you!

Reply:
Slight update after looking at what Semaphore does with its secrets here: https://github.com/semaphore-protocol/semaphore/blob/v3.15.2/packages/identity/src/identity.ts#L22

- Identity and trapdoor are the raw random inputs, which are the "private key" in V3, so I think my proposed use of them is reasonable.
- I suggest putting them in the order `nullifier` then `trapdoor` (opposite of the order I proposed above) to be consistent with how Semaphore does its hashing and naming.
- I didn't remember the role of the `secret` being an extra layer of hashing before the `commitment`.  I'm not 100% sure what it's used for.  We could use just the `secret` as the V4 private key as-is (as a 32-byte buffer), which would be simpler code, but I feel like the implicit use of a Poseidon hash here seems wrong when we're doing something non-ZK-friendly anyway.
- We might also consider using a blake hash instead of sha256 to match EdDSA-Poseidon, which uses a 512-bit blake1 hash.  Not sure if that's worthwhile or not.

I'm going to run the ideas above past Cedoor in case he has thoughts from the Semaphore side.  Meta-commentary for you is to think about when we'll cross the threshold of these migrations being "locked in" for real users, and make sure we finalize the decision of hash function before then.  Until then it can be fungible.